src/compiler/scanner.ts

```diff
@@ -1101,6 +1100,40 @@ export function createScanner(languageVersion: ScriptTarget, skipTrivia: boolean
 
     return scanner;
 
+    /**
+     * Returns the code point for the character at the given position within `text`. This
+     * should only be used when pos is guaranteed to be within the bounds of `text` as this
+     * function does not perform bounds checks.
+     */
+    function codePointUnchecked(pos: number) {
+        return codePointAt(text, pos);

```

Comment:
`codePointAt` is also just a little wrapper that pretends that it's always in bounds; maybe that's something to fix up in the follow-up you mentioned.

Reply:
`text.codePointAt` returns `null` for an out of bounds index, which is just as bad as `NaN`. I can probably remove the intermediate call, but it's inlined anyways so it won't affect perf.

Reply:
Yeah, it was the `null` return that I was thinking about; with this PR applied, debug asserting on out of bounds calls to `codePointAt` fails a few hundred tests I thought would be outside of this PR from other callers (like skipTrivia), though I was surprised to see some code paths which were in the new code:

```
  45) 
       
         conformance tests
           conformance tests for tests/cases/conformance/parser/ecmascript5/SkippedTokens/parserSkippedTokens18.ts
             "before all" hook for "Correct errors for tests/cases/conformance/parser/ecmascript5/SkippedTokens/parserSkippedTokens18.ts":
     Error: Debug Failure. False expression: Invalid index
      at codePointAt (src/compiler/scanner.ts:3922:11)
      at codePointUnchecked (src/compiler/scanner.ts:1109:16)
      at peekExtendedUnicodeEscape (src/compiler/scanner.ts:1730:13)
      at Object.scan (src/compiler/scanner.ts:2264:48)


  285) 
       
         unittests:: tsserver:: applyChangesToOpenFiles
           with updateOpen request:
     Error: Debug Failure. False expression: Invalid index
      at codePointAt (src/compiler/scanner.ts:3922:11)
      at codePointUnchecked (src/compiler/scanner.ts:1109:16)
      at checkForIdentifierStartAfterNumericLiteral (src/compiler/scanner.ts:1304:32)
      at scanNumber (src/compiler/scanner.ts:1298:13)
      at Object.scan (src/compiler/scanner.ts:2134:36)
```

Maybe fixing these would help perf some more too?

Reply:
```diff
diff --git a/src/compiler/scanner.ts b/src/compiler/scanner.ts
index 1fb0c63024..3446178dc9 100644
--- a/src/compiler/scanner.ts
+++ b/src/compiler/scanner.ts
@@ -983,6 +983,8 @@ export function isIdentifierPart(ch: number, languageVersion: ScriptTarget | und
 
 /** @internal */
 export function isIdentifierText(name: string, languageVersion: ScriptTarget | undefined, identifierVariant?: LanguageVariant): boolean {
+    if (name === "") return false;
+
     let ch = codePointAt(name, 0);
     if (!isIdentifierStart(ch, languageVersion)) {
         return false;
@@ -1301,7 +1303,7 @@ export function createScanner(languageVersion: ScriptTarget, skipTrivia: boolean
     }
 
     function checkForIdentifierStartAfterNumericLiteral(numericStart: number, isScientific?: boolean) {
-        if (!isIdentifierStart(codePointUnchecked(pos), languageVersion)) {
+        if (!isIdentifierStart(codePointChecked(pos), languageVersion)) {
             return;
         }
 
@@ -1727,7 +1729,7 @@ export function createScanner(languageVersion: ScriptTarget, skipTrivia: boolean
     }
 
     function peekExtendedUnicodeEscape(): number {
-        if (codePointUnchecked(pos + 1) === CharacterCodes.u && codePointUnchecked(pos + 2) === CharacterCodes.openBrace) {
+        if (codePointChecked(pos + 2) === CharacterCodes.openBrace && codePointUnchecked(pos + 1) === CharacterCodes.u) {
             const start = pos;
             pos += 3;
             const escapedValueString = scanMinimumNumberOfHexDigits(1, /*canHaveSeparators*/ false);
@@ -2945,7 +2947,7 @@ export function createScanner(languageVersion: ScriptTarget, skipTrivia: boolean
                         error(Diagnostics.A_character_class_range_must_not_be_bounded_by_another_character_class, maxStart, pos - maxStart);
                         continue;
                     }
-                    if (!minCharacter) {
+                    if (!minCharacter || !maxCharacter) {
                         continue;
                     }
                     const minCharacterValue = codePointAt(minCharacter, 0);
@@ -3919,6 +3921,7 @@ export function createScanner(languageVersion: ScriptTarget, skipTrivia: boolean
 
 function codePointAt(s: string, i: number): number {
     // TODO(jakebailey): this is wrong and should have ?? 0; but all users are okay with it
+    Debug.assert(i >= 0 && i < s.length, "Invalid index");
     return s.codePointAt(i)!;
 }
 
```

Is what I got that made this particular function pass.

Reply:
If I do the same assert in both `codePointUnchecked` and `charCodeUnchecked` we end up with nearly 20k test failures, so we're definitely not performing accurate bounds checks during scan.